### PR TITLE
LibCore: Refactor template definitions to header in ArgsParser

### PR DESCRIPTION
**Pull request by Ollrogge**
This option is useful for parsing e.g. negative offsets. Wondering why this was missing but it works fine :)

### CODE REVIEW

File: Userland/Libraries/LibCore/ArgsParser.cpp [hunk 1]
```diff
@@ -595,6 +595,49 @@ void ArgsParser::add_option(Vector<size_t>& values, char const* help_string, cha
...
+        OptionArgumentMode::Required,
+        help_string,
+        long_name,
+        short_name,
+        value_name,
+        [&values, separator](StringView s) -> ErrorOr<bool> {
+            bool parsed_all_values = true;
+
+            s.for_each_split_view(separator, SplitBehavior::Nothing, [&](auto value) {
+                if (auto maybe_value = AK::StringUtils::convert_to_uint<ssize_t>(value); maybe_value.has_value())
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibCore/ArgsParser.cpp:627]: As noted by the CI failure, you can't use `convert_to_uint` for signed ints. `value.to_number<ssize_t>()` should be fine I think.

Comment (Ollrogge) [reply Userland/Libraries/LibCore/ArgsParser.cpp]: I used convert_to_int now

**Comment by timschumi** [comment]
Absolutely tiniest POSIX-based nitpick: `ssize_t` is only guaranteed to have the range `[-1, SSIZE_MAX]`, so it technically isn't fit for any kind of offset. If one needs actual offsets, there is `off_t`.

**Comment by trflynn89** [comment]
> Wondering why this was missing 

The reason generally is we don't add code that isn't used. Do you have a use case for this option in mind?

**Comment by Ollrogge** [comment]
> > Wondering why this was missing
> 
> The reason generally is we don't add code that isn't used. Do you have a use case for this option in mind?

Yes, I am currently implementing xxd which has an option (-s) to seek to a position in a file. This seek position can be relative to the end of the file, therefore I needed a way to parse negative numbers.

I didn't know however that `ssize_t` is only guaranteed to be in the of range `[-1, SSIZE_MAX]`. So I guess I should be using `off_t` or rather `i64` in this case which is already implemented by ArgsParser.

Maybe adding another template instantiation with `off_t` as parameter to make the intention of using an offset clear ?

**Comment by AtkinsSJ** [comment]
Could we instead have templated "any integer" overloads, instead of adding them for each int type we want? eg:

```c++
template <Integral T>
void add_option(Optional<T>& value, char const* help_string, char const* long_name, char short_name, char const* value_name, OptionHideMode hide_mode = OptionHideMode::None);
template <Integral T>
void add_option(Vector<T>& values, char const* help_string, char const* long_name, char short_name, char const* value_name, char separator = ',', OptionHideMode hide_mode = OptionHideMode::None);
```

Those would need to use `string_view.to_number<T>();`, and would need to be directly in the header. But maybe more generally useful.

**Comment by Ollrogge** [comment]
> Could we instead have templated "any integer" overloads, instead of adding them for each int type we want? eg:
> 
> ```c++
> template <Integral T>
> void add_option(Optional<T>& value, char const* help_string, char const* long_name, char short_name, char const* value_name, OptionHideMode hide_mode = OptionHideMode::None);
> template <Integral T>
> void add_option(Vector<T>& values, char const* help_string, char const* long_name, char short_name, char const* value_name, char separator = ',', OptionHideMode hide_mode = OptionHideMode::None);
> ```
> 
> Those would need to use `string_view.to_number<T>();`, and would need to be directly in the header. But maybe more generally useful.

Following your suggestion, I have refactored the integral template functions to be in the header. Seems to work nicely !

**Comment by Ollrogge** [comment]
Any other changes required ?

**Review by AtkinsSJ — APPROVED** [review]
Looks good! I'd just forgotten about it. :sweat_smile: Thanks for doing this!

**Comment by Ollrogge** [comment]
Thank you ! :)

**Pull request merged**